tensorflow/core/kernels/mkl_relu_op.cc

```diff
@@ -0,0 +1,390 @@
+
+    relu_res[dnnResourceDst] = user_o;
+    relu_res[dnnResourceSrc] = user_i;
+    CHECK_EQ(dnnExecute_F32(mkl_prim_relu_fwd_, relu_res), E_SUCCESS);
+    Mklcleanup();
+  }
+
+ private:
+  typedef struct {
+    int in_dims;
+    size_t *in_sizes;
+    size_t *in_strides;
+    MklShape input_shape, output_shape;
+  } MklReluOpParams_;
+
+  void Mklcleanup() {
+    bool input_in_mkl_format = mkl_params.input_shape.IsMklTensor();
+    if (!input_in_mkl_format) dnnLayoutDelete_F32(mkl_lt_input_);
+    dnnDelete_F32(mkl_prim_relu_fwd_);
+  }
+  void MklCreateInputLayouts(OpKernelContext* context) {
+    bool input_in_mkl_format = mkl_params.input_shape.IsMklTensor();
+    if (!input_in_mkl_format) {
+      CHECK_EQ(dnnLayoutCreate_F32(&mkl_lt_input_,
+          mkl_params.in_dims, mkl_params.in_sizes,
+                            mkl_params.in_strides),
+               E_SUCCESS);
+    } else {
+      mkl_lt_input_ = static_cast<dnnLayout_t>
+          (mkl_params.input_shape.GetCurLayout());
+    }
+  }
+
+  dnnPrimitive_t mkl_prim_relu_fwd_ = nullptr;
+  MklReluOpParams_ mkl_params;
+  void* relu_res[dnnResourceNumber];
+  dnnLayout_t mkl_lt_input_ = nullptr;
+};
+
+template <typename Device, typename T>
+class MklReluGradOp : public OpKernel {
+ public:
+  ~MklReluGradOp() {}
+  explicit MklReluGradOp(OpKernelConstruction* context) : OpKernel(context) {}
+  void Compute(OpKernelContext* context) override;
+
+ private:
+  typedef struct {
+    int in_dims;
+    size_t *in_sizes;
+    size_t *in_strides;
+    MklShape input_shape, grad_shape, output_shape;
+  } MklReluGradOpParams_;
+  MklReluGradOpParams_ mkl_params;
+
+  void MklPrepareReluGradInputs(OpKernelContext* context,
+                                   Tensor* mkl_tmp_grad_buf_tensor,
+                                   Tensor* mkl_tmp_input_buf_tensor) {
+  dnnPrimitive_t cv_user_to_reluB_input = nullptr,
+      cv_user_to_reluB_grad = nullptr;
+  dnnLayout_t mkl_lt_internal_input = nullptr, mkl_lt_internal_grad = nullptr;
+
+  const Tensor& g = MklGetInput(context, 0);
+  const Tensor& a = MklGetInput(context, 1);
+
+  void* user_i = static_cast<void*>(const_cast<T*>(a.flat<T>().data()));
+  void* user_g = static_cast<void*>(const_cast<T*>(g.flat<T>().data()));
+
+  CHECK_EQ(dnnLayoutCreateFromPrimitive_F32(&mkl_lt_internal_grad,
+      mkl_prim_relu_back_,
+     dnnResourceDiffDst),
+           E_SUCCESS);
+
+  CHECK_EQ(dnnLayoutCreateFromPrimitive_F32(&mkl_lt_internal_input,
+      mkl_prim_relu_back_,
+      dnnResourceSrc),
+           E_SUCCESS);
+
+  if (!dnnLayoutCompare_F32(mkl_lt_internal_grad, mkl_lt_grad_)) {
+    AllocTmpBuffer(context, mkl_tmp_grad_buf_tensor, mkl_lt_internal_grad,
+                   &relu_res[dnnResourceDiffDst]);
+    CHECK_EQ(dnnConversionCreate_F32(&cv_user_to_reluB_grad, mkl_lt_grad_,
+                                     mkl_lt_internal_grad),
+             E_SUCCESS);
+  }
+
+  if (!dnnLayoutCompare_F32(mkl_lt_internal_input, mkl_lt_input_)) {

```

Comment:
These parameter names are very cryptic. If you can't make them more descriptive, then please add comments explaining each parameter.